src/agents/graph.py

```diff
@@ -0,0 +1,342 @@
+            )
+            state.subtasks = self.parser.parse(plan.content).subtasks
+            return {
+                "subtasks": state.subtasks,
+                "messages": (
+                    [
+                        AIMessage(
+                            content=f"Task decomposed into subtasks and assigned to agents: {state.subtasks}",
+                            name=PLANNER,
+                        )
+                    ]
+                ),
+                "error": None,
+            }
+        except Exception as e:
+            logger.error(f"Error in planning: {e}")
+            return {
+                "error": str(e),
+                "messages": [
+                    AIMessage(
+                        content=f"Error in planning: {e}",
+                        name="Planner",
+                    )
+                ],
+            }
+
+    def common_node(self, state: AgentState) -> dict[str, Any]:
+        """Breaks down the given user query into sub-tasks."""
+
+        prompt = ChatPromptTemplate.from_messages(
+            [
+                (
+                    "system",
+                    "Given the conversation and the last user query, "
+                    "you are tasked with generating a response. ",
+                ),
+                MessagesPlaceholder(variable_name="messages"),
+                ("human", "query: {query}"),
+            ]
+        )
+        chain = prompt | self.model.llm
+
+        for subtask in state.subtasks:
+            if subtask.assigned_to == COMMON and subtask.status != "completed":
+                try:
+                    response = chain.invoke(
+                        {
+                            "messages": state.messages,
+                            "query": subtask.description,
+                        },
+                        config={"callbacks": [handler]},
+                    ).content
+                    subtask.complete()
+                    return {
+                        "messages": [
+                            AIMessage(
+                                content=response,
+                                name=COMMON,
+                            )
+                        ],
+                    }
+                except Exception as e:
+                    logger.error(f"Error in common node: {e}")
+                    return {
+                        "error": str(e),
+                        "messages": [
+                            AIMessage(
+                                content=f"Error occurred: {e}",
+                                name=COMMON,
+                            )
+                        ],
+                    }
+        return {
+            "messages": [
+                AIMessage(
+                    content="All my subtasks are already completed.",
+                    name=COMMON,
+                )
+            ]
+        }
+
+    def generate_final_response(self, state: AgentState) -> dict[str, Any]:
```

Comment:
Shall we introduce final_response as a separate agent?

Reply:
if you ask my opinion, I wouldn't do it as the complex work is already done by the specific agents. 
I'll document the point in the issue so that we think together. Is it okay?